### PR TITLE
feat(profiling): integrate scoped-based timer to onedim core

### PR DESCRIPTION
**Pull request by tai271828**
This pull request is the first step of #135 .

# How to test this pull request
Run this driving script from gist [euler1d_profiling-py](https://gist.github.com/tai271828/d5f50e4a06274b1203ed79f6739039dd#file-euler1d_profiling-py). The following output will be generated on your stdout after all pictures of shock tube are played and the image window is closed.

```
Euler1DCore.march_alpha2 : count = 51 , time = 0.0365707 (second)
Euler1DCore.setup_march : count = 1 , time = 2.2676e-05 (second)
Euler1DCore::density : count = 51 , time = 0.0020337 (second)
Euler1DCore::initialize_data : count = 1 , time = 1.6329e-05 (second)
Euler1DCore::march_half_so0 : count = 102 , time = 0.0179795 (second)
Euler1DCore::pressure : count = 51 , time = 0.00219614 (second)
Euler1DCore::update_cfl : count = 103 , time = 0.00225837 (second)
Euler1DCore::velocity : count = 51 , time = 0.00142665 (second)
Euler1DKernel::derive : count = 20298 , time = 0.0137601 (second)
Euler1DKernel::march_half_so1_alpha : count = 102 , time = 0.01569 (second)
```

### CODE REVIEW

File: cpp/modmesh/onedim/Euler1DCore.cpp [hunk 1]
```diff
@@ -111,6 +117,7 @@ void Euler1DCore::update_cfl(bool odd_plane)
 
 void Euler1DCore::march_half_so0(bool odd_plane)
```
Comment (yungyuc) [inline cpp/modmesh/onedim/Euler1DCore.cpp:118]: Could you please also add the profiling hooks to the first-order derivative calculating functions (templates)?

**Review by yungyuc — CHANGES_REQUESTED** [review]
Could you also add the profiling driver in a gist and include it in the PR for reference?

File: cpp/modmesh/onedim/Euler1DCore.cpp (hunk 1 above)
Comment (tai271828) [reply cpp/modmesh/onedim/Euler1DCore.cpp:118]: Will do!

Comment (tai271828) [reply cpp/modmesh/onedim/Euler1DCore.cpp:118]: The gist has been appended in the original description.

Comment (tai271828) [reply cpp/modmesh/onedim/Euler1DCore.cpp:118]: Completed and updated this PR.

**Comment by tai271828** [comment]
I prepared a new driving script from gist [euler1d_profiling-py](https://gist.github.com/tai271828/d5f50e4a06274b1203ed79f6739039dd#file-euler1d_profiling-py) for testing (and also updated the corresponding description session above). This driving script runs through a real shock tube problem and the result is easier to ready.

**Comment by yungyuc** [comment]
> I prepared a new driving script from gist [euler1d_profiling-py](https://gist.github.com/tai271828/d5f50e4a06274b1203ed79f6739039dd#file-euler1d_profiling-py) for testing (and also updated the corresponding description session above). This driving script runs through a real shock tube problem and the result is easier to ready.

What about integrating the script `euler1d_profiling.py` to the euler1d app and then simplifying the driving script to be a one-liner?  The one-liner can be mentioned in the comment in the app.

**Comment by tai271828** [comment]
> > I prepared a new driving script from gist [euler1d_profiling-py](https://gist.github.com/tai271828/d5f50e4a06274b1203ed79f6739039dd#file-euler1d_profiling-py) for testing (and also updated the corresponding description session above). This driving script runs through a real shock tube problem and the result is easier to ready.
> 
> What about integrating the script `euler1d_profiling.py` to the euler1d app and then simplifying the driving script to be a one-liner? The one-liner can be mentioned in the comment in the app.

Sounds good. Let me work on it.

File: modmesh/app/euler1d.py [hunk 2]
```diff
@@ -25,6 +25,9 @@
 # POSSIBILITY OF SUCH DAMAGE.
 
 
+__package__ = "modmesh.app"
```
Comment (tai271828) [inline modmesh/app/euler1d.py:29]: A trick to make euler.py runnable in command line.

File: modmesh/app/euler1d.py [hunk 3]
```diff
@@ -46,6 +49,9 @@ def load_app():
 # Use the functions for more examples:
 ctrl.start()  # Start the movie
 ctrl.step()  # Stepping the solution
+
+# Note: you can enable profiling info by "profiling" option
+ctrl = mm.app.euler1d.run(interval=10, max_steps=50, profiling=True)
```
Comment (tai271828) [inline modmesh/app/euler1d.py:54]: The one-liner aforementioned https://github.com/solvcon/modmesh/pull/190#issuecomment-1383108362 .

File: modmesh/app/euler1d.py [hunk 4]
```diff
@@ -175,16 +185,56 @@ def log(msg):
...
+
+def run(interval=10, max_steps=50, no_view_mgr=False, **kw):
     st = euler1d.ShockTube()
     st.build_constant(gamma=1.4, pressure1=1.0, density1=1.0, pressure5=0.1,
                       density5=0.125)
     st.build_numerical(xmin=-10, xmax=10, ncoord=201, time_increment=0.05)
 
-    ctrl = Controller(shocktube=st, max_steps=max_steps, **kw)
+    if no_view_mgr:
+        ctrl = ControllerNoViewMgr(shocktube=st, max_steps=max_steps, **kw)
```
Comment (tai271828) [inline modmesh/app/euler1d.py:208]: When running in a terminal (for automation), we don't launch view manager.

File: modmesh/app/euler1d.py (hunk 3 above)
Comment (yungyuc) [reply modmesh/app/euler1d.py:54]: This is great!

**Review by yungyuc — APPROVED** [review]

File: modmesh/app/euler1d.py (hunk 2 above)
Comment (tai271828) [reply modmesh/app/euler1d.py:29]: @yungyuc thanks for adding the comments. This one is really tricky so I am also very +1 to adding comments on it. Thanks for the update!

**Pull request merged**